dbms/src/Operators/HashProbeTransformExec.cpp

```diff
@@ -0,0 +1,259 @@
+
+            return restore_probe_exec;
+        }
+        assert(join->hasPartitionSpilledWithLock() == false);
+    }
+
+    // current join has no more partition to restore, so check if previous join still has partition to restore
+    return parent ? parent->tryGetRestoreExec() : HashProbeTransformExecPtr{};
+}
+
+void HashProbeTransformExec::startRestoreProbe()
+{
+    /// Trigger probe side restore task to get the block from probe_restore_stream.
+    ///
+    /// StreamRestoreTask [probe_restore_stream]
+    ///                       | read and push
+    ///                       ▼
+    ///               probe_result_queue
+    ///                       | pop and probe
+    ///                       ▼
+    ///             HashProbeTransformExec (probe restored hash partition)
+    assert(!is_probe_restore_done && probe_restore_stream);
+    // Use 1 as the queue_size to avoid accumulating too many blocks and causing the memory to exceed the limit.
+    assert(!probe_result_queue);
+    probe_result_queue = std::make_shared<ResultQueue>(1);
+    TaskScheduler::instance->submit(std::make_unique<StreamRestoreTask>(exec_context, log->identifier(), probe_restore_stream, probe_result_queue));
+    probe_restore_stream.reset();
+}
+
+bool HashProbeTransformExec::prepareProbeRestoredBlock()
+{
+    if (unlikely(is_probe_restore_done))
+        return true;
+    if (probe_restored_block)
+        return true;
+    assert(probe_result_queue);
+    auto ret = probe_result_queue->tryPop(probe_restored_block);
+    switch (ret)
+    {
+    case MPMCQueueResult::OK:
+        return true;
+    case MPMCQueueResult::EMPTY:
+        return false;
+    case MPMCQueueResult::FINISHED:
+        is_probe_restore_done = true;
+        return true;
+    default:
+        throw Exception(fmt::format("Unexpected result: {}", magic_enum::enum_name(ret)));
+    }
+}
+
+Block HashProbeTransformExec::popProbeRestoredBlock()
+{
+    if (unlikely(is_probe_restore_done))
+        return {};
+
+    // `prepareProbeRestoredBlock` must have been called and return true before,
+    // so `probe_restored_block` must not be empty here.
+    Block ret;
+    assert(probe_restored_block);
+    std::swap(ret, probe_restored_block);
+    assert(!probe_restored_block);
+    return ret;
+}
+
+#define CONTINUE                             \
+    if unlikely (exec_context.isCancelled()) \
+        return OperatorStatus::CANCELLED;    \
+    continue;
+
+OperatorStatus HashProbeTransformExec::tryFillProcessInfoInRestoreProbeStage(ProbeProcessInfo & probe_process_info)
+{
+    while (true)
+    {
+        assert(probe_process_info.all_rows_joined_finish);
+        if (!probe_partition_blocks.empty())
+        {
+            auto partition_block = std::move(probe_partition_blocks.front());
+            probe_partition_blocks.pop_front();
+            if (!partition_block.block || partition_block.block.rows() == 0)
+            {
+                CONTINUE;
```

Comment:
Ditto

Reply:
Because it is in a while loop, here we check `context.isCancelled` in each loop of while to avoid canceling too slowly.